plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Service.java

```diff
@@ -210,27 +241,108 @@ private synchronized void releaseCachedClients() {
         IdleConnectionReaper.shutdown();
     }
 
-    static class PrivilegedInstanceProfileCredentialsProvider implements AWSCredentialsProvider {
-        private final AWSCredentialsProvider credentials;
+    @Override
+    public void close() {
+        releaseCachedClients();
+        WEB_IDENTITY_TOKEN_CREDENTIALS_PROVIDER.shutdown();
+    }
+
+    static class PrivilegedAWSCredentialsProvider implements AWSCredentialsProvider {
+        private final AWSCredentialsProvider credentialsProvider;
+
+        private PrivilegedAWSCredentialsProvider(AWSCredentialsProvider credentialsProvider) {
+            this.credentialsProvider = credentialsProvider;
+        }
 
-        private PrivilegedInstanceProfileCredentialsProvider() {
-            // InstanceProfileCredentialsProvider as last item of chain
-            this.credentials = new EC2ContainerCredentialsProviderWrapper();
+        AWSCredentialsProvider getCredentialsProvider() {
+            return credentialsProvider;
         }
 
         @Override
         public AWSCredentials getCredentials() {
-            return SocketAccess.doPrivileged(credentials::getCredentials);
+            return SocketAccess.doPrivileged(credentialsProvider::getCredentials);
         }
 
         @Override
         public void refresh() {
-            SocketAccess.doPrivilegedVoid(credentials::refresh);
+            SocketAccess.doPrivilegedVoid(credentialsProvider::refresh);
         }
     }
 
-    @Override
-    public void close() {
-        releaseCachedClients();
+    /**
+     * Customizes {@link com.amazonaws.auth.WebIdentityTokenCredentialsProvider}
+     *
+     * <ul>
+     * <li>Reads the the location of the web identity token not from AWS_WEB_IDENTITY_TOKEN_FILE, but from a symlink
+     * in the plugin directory, so we don't need to create a hardcoded read file permission for the plugin.</li>
+     * <li>Supports customization of the STS endpoint via a system property, so we can test it against a test fixture.</li>
+     * <li>Supports gracefully shutting down the provider and the STS client.</li>
+     * </ul>
+     */
+    static class CustomWebIdentityTokenCredentialsProvider implements AWSCredentialsProvider {
+
+        private STSAssumeRoleWithWebIdentitySessionCredentialsProvider credentialsProvider;
+        private AWSSecurityTokenService stsClient;
+
+        @SuppressForbidden(reason = "Need to use PathUtils#get")
+        CustomWebIdentityTokenCredentialsProvider() {
+            // Check whether the original environment variable exists. If it doesn't,
+            // the system doesn't support AWS web identity tokens
+            if (System.getenv(AWS_WEB_IDENTITY_ENV_VAR) == null) {
+                return;
+            }
+            String esPathConf = System.getProperty("es.path.conf");
```

Comment:
We should retrieve the plugin's configuration directory from the `Environment` instead of relying on the sysprop. Maybe you can pass it down from the `Plugin` to the S3Service instance.

Reply:
Yeah, that was my original idea, but I couldn't figure out hot to pass `Environment` all the way down from `Plugin`

Reply:
I think that the S3Service could be created as a component in S3RepositoryPlugin and later reused in the repository factory? See SearchableSnapshot.createComponents() method and the FrozenCacheService instance which is later reused in SearchableSnapshot.getDirectoryFactories()

Reply:
Done!